### ISSUE TITLE
Digitally sign openrefine.exe to eliminate extra security warnings on Windows

### ISSUE THREAD

**Issue by thadguidry** [OpenRefine/OpenRefine#3003]
**Is your feature request related to a problem or area of OpenRefine? Please describe.**
Currently on Windows 10, clicking `openrefine.exe` will cause extra security warnings for Microsoft Defender SmartScreen.

**Describe the solution you'd like**
I think only 1 option exists:

1. Since OpenRefine is downloaded from GitHub website and not through our own OpenRefine website the SmartScreen technology considers the openrefine.exe as unsafe since it is not digitally signed and verified with Microsoft.

**Describe alternatives you've considered**
* Bypassing SmartScreen every time to run - must click "More Info" on SmartScreen blocking dialog and then "Run Anyways"
* Bypassing SmartScreen permanently - After downloading and extracting the .zip file, right clicking on `openrefine.exe` and clicking **Unblock** to add openrefine.exe to safe list for SmartScreen so that it does not block and consider it safe.

**Additional context**
Some non-exhaustive details I have found:
* https://fb.smartscreen.microsoft.com/smartscreenfaq.aspx
* [Windows App Certification Kit](https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/windows-app-certification-kit)
* https://answers.microsoft.com/en-us/protect/forum/all/how-verifyed-publisher-status-makes-an-application/d4c81d5c-79bf-4caf-89da-5e4747b571e7

**Comment by rybesh** [comment]
This would be a big help for students on locked-down school laptops.

**Comment by georgiawebgurl** [comment]
My new PC laptop / Windows 10 installed both Microsoft Defender and Mcafee. I've had an extremely hard time getting OpenRefine 3.4.1 to run. I've tried adding it as safe, running it as admin, etc. It kept throwing the error that it can't create the workspace. I eventually figured out how to work around the issue by going through Windows at the system level to change the program settings. However, I lost my custom settings in the last Window upgrade. I also teach OpenRefine workshops and provide metadata/training support to a variety of institutions. This is a huge deal for many of us, not just students. Many metadata workers, librarians, archivists, data analysts, etc. use OpenRefine. I am more than happy to help in testing, as I have tim because this is a really a huge issue for my community. Thanks!

**Comment by wetneb** [comment]
@georgiawebgurl thank you *so much* for this feedback and your offer to help with testing. This is really helpful. I am acutely aware of the importance of improving OpenRefine's packaging, but not being personally exposed to the problem much on a day to day basis I really need reminders like yours that this is really critical.

I hope to follow up soon on your offer.

**Comment by georgiawebgurl** [comment]
Sounds great! Thank you! :-)

**Comment by georgiawebgurl** [comment]
I just want to report that with the most recent Windows Upgrade to 
Windows MS 10 Home Build 10.0.19041 build 19041, 'Run as Administrator' and then accepting the security prompt, does allow OpenRefine to work without issue.

**Comment by wetneb** [comment]
By the way the OpenRefine project has funding available to work on this. Do you know someone who would be able to get this fixed? Send them our way!
https://openrefine.org/blog/2022/09/30/windows-macos-packaging.html

**Comment by wetneb** [comment]
Some more ideas to solve this issue: there is [a standard procedure to sign executable files generated by launch4j](https://sourceforge.net/p/launch4j/discussion/332684/thread/d4985f08/). It relies on [jsign](https://ebourg.github.io/jsign/), a Java implementation of Windows' signing protocol.

**Comment by wetneb** [comment]
So if I understand correctly, this would only require us buying a code signing certificate from one of the many vendors available (not Microsoft directly), and then we should be able to do the signing in a relatively straightforward way in the CI (even in our current MacOS job which is in charge of the packaging of all versions).

**Comment by thadguidry** [comment]
That's my understanding, however not sure on the keystore type needed.  I guess it would matter security wise, if a GitHub runner was accidentally compromised and stayed alive?
Hmm, also the keypass could be used, but how would that be shared privately and ensure only a select few maintainers have it and type it in during release building?

**Comment by wetneb** [comment]
We have the exact same problems with the MacOS signing certificate. Access to that is granted by GitHub secrets that are available in the CI that builds releases (but not PRs). For the MacOS certificate, access to it is managed by CS&S (they can add Apple ID accounts of maintainers to the system).
So I do not think this should be a significant hurdle.

**Comment by jas88** [comment]
> So if I understand correctly, this would only require us buying a code signing certificate from one of the many vendors available (not Microsoft directly), and then we should be able to do the signing in a relatively straightforward way in the CI (even in our current MacOS job which is in charge of the packaging of all versions).

That's right - buy a code-signing certificate (that's the difficult/expensive bit IME, as they verify the buying company's identity - something like $500 a year to Digicert), put the private key part in a Github 'secret', then write a few lines of Github Action to retrieve that into a .pfx file and run it through signtool.exe. I recently updated our signing script for an open source project of ours: https://github.com/HicServices/DicomTemplateBuilder/blob/main/.github/workflows/dotnet-desktop.yml
